### PR TITLE
Scatter resources

### PR DESCRIPTION
**Pull request by proger**
The main idea is to make resources hackable separately. I'm unsure about separating functions in `Types/Resources.h` from the instance definitions, however that avoids orphans and makes functions usable separately.

While here, haddockify some comments and bump to `lts-5.3`.

Ping @fmap

### CODE REVIEW

File: src/Upcast/Infra/Types/Amazonka.hs [hunk 1]
```diff
@@ -28,6 +30,7 @@ readRegion s = case unpack s of
   "us-gov-west-1"      -> GovCloud
   "fips-us-gov-west-1" -> GovCloudFIPS
   "sa-east-1"          -> SaoPaulo
+  unk -> error ("unknown region " ++ unk)
```
Comment (fmap) [inline src/Upcast/Infra/Types/Amazonka.hs]: indent

File: src/Upcast/Infra/Types/Resource.hs [hunk 2]
```diff
@@ -150,512 +111,111 @@ class Resource a where
...
-    EC2.importKeyPair name pubkey & send <&> view (EC2.ikprsKeyName . _Just)
-  update :: AWS m => ResourceId -> Ec2keypair -> m ResourceId
-  update current Ec2keypair{..} = do
-    keypair <- create Ec2keypair{..}
-    EC2.deleteKeyPair current & defer . send
-    return keypair
+  create = Keypair.create
+  update = Keypair.update
+
   match Ec2keypair{..} = do
```
Comment (fmap) [inline src/Upcast/Infra/Types/Resource.hs]: Why not scatter this?

**Comment by fmap** [comment]
>  I'm unsure about separating functions in Types/Resources.h from the instance definitions,

The lack of context seems awkward, yes.

> however that avoids orphans and makes functions usable separately.

I agree.

Modulo comments, :shipit:.

File: src/Upcast/Infra/Types/Resource.hs (hunk 2 above)
Comment (proger) [inline src/Upcast/Infra/Types/Resource.hs]: It seems that bit of state has nothing to do with the actual "cloud ops", so i left the state code there.

Comment (fmap) [inline src/Upcast/Infra/Types/Resource.hs]: :+1:

**Pull request merged**